cpp/src/reshape/table_to_array.cu

```diff
@@ -0,0 +1,133 @@
+
+template <typename T>
+void table_to_array_iml(table_view const& input, void* output, rmm::cuda_stream_view stream)
+{
+  auto const num_columns = input.num_columns();
+  auto const num_rows    = input.num_rows();
+  auto const item_size   = sizeof(T);
+  auto* base_ptr         = static_cast<cuda::std::byte*>(output);
+
+  std::vector<void*> h_srcs(num_columns);
+  std::vector<void*> h_dsts(num_columns);
+
+  for (int i = 0; i < num_columns; ++i) {
+    auto const& col = input.column(i);
+    CUDF_EXPECTS(col.type() == input.column(0).type(), "All columns must have the same dtype");
+    CUDF_EXPECTS(col.null_count() == 0, "All columns must be non-nullable or contain no nulls");
+
+    h_srcs[i] = const_cast<void*>(static_cast<void const*>(col.data<T>()));
+    h_dsts[i] = static_cast<void*>(base_ptr + i * item_size * num_rows);
+  }
+
+  auto const mr = cudf::get_current_device_resource_ref();
+
+  auto d_srcs = cudf::detail::make_device_uvector_async(h_srcs, stream, mr);
+  auto d_dsts = cudf::detail::make_device_uvector_async(h_dsts, stream, mr);
+
+  thrust::constant_iterator<size_t> sizes(static_cast<size_t>(item_size * num_rows));
+
+  void* d_temp_storage      = nullptr;
+  size_t temp_storage_bytes = 0;
+  cub::DeviceMemcpy::Batched(d_temp_storage,
+                             temp_storage_bytes,
+                             d_srcs.begin(),
+                             d_dsts.begin(),
+                             sizes,
+                             num_columns,
+                             stream.value());
+
+  rmm::device_buffer temp_storage(temp_storage_bytes, stream);
+
+  cub::DeviceMemcpy::Batched(temp_storage.data(),
+                             temp_storage_bytes,
+                             d_srcs.begin(),
+                             d_dsts.begin(),
+                             sizes,
+                             num_columns,
+                             stream.value());
+}
+
+struct table_to_array_dispatcher {
+  table_view const& input;
+  void* output;
+  rmm::cuda_stream_view stream;
+
+  template <typename T, CUDF_ENABLE_IF(is_fixed_width<T>())>
+  void operator()() const
+  {
+    table_to_array_iml<T>(input, output, stream);
+  }
+
+  template <typename T, CUDF_ENABLE_IF(!is_fixed_width<T>())>
+  void operator()() const
+  {
+    CUDF_FAIL("Unsupported dtype");
+  }
+};
+
+}  // namespace
+
+void table_to_array(table_view const& input,
+                    void* output,
+                    data_type output_dtype,
+                    rmm::cuda_stream_view stream)
+{
+  CUDF_EXPECTS(output != nullptr, "Output pointer cannot be null.", std::invalid_argument);
+  CUDF_EXPECTS(
+    input.num_columns() > 0, "Input must have at least one column.", std::invalid_argument);
+
+  cudf::type_dispatcher<cudf::dispatch_storage_type>(
+    output_dtype, table_to_array_dispatcher{input, output, stream});
+}
+
+}  // namespace detail
+
+void table_to_array(table_view const& input,
+                    void* output,
```

Comment:
I wonder if we should make this `output` pointer into a span of some kind, or require a size parameter. Having a "sized view" might make it easier to ensure that we aren't writing out of bounds for the user-provided pointer.

Reply:
Done in https://github.com/rapidsai/cudf/pull/18450/commits/82c5b22de5540c85770542c49cfebc92f6fc82ad